### ISSUE TITLE
[Bug]: The engine died and cant be restarted

### ISSUE THREAD

**Issue by angelapytao** [towhee-io/towhee#1651]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

```
In multiprocessing on CentOS,in sub process cannot start the engine, but on MacOS, the sub code is running OK.
torch.device("cuda" if torch.cuda.is_available() else "cpu")
pp = pipeline('towhee/image-embedding-resnet50')

### Expected Behavior

on Centos in sub process pipeline run.

### Steps To Reproduce

_No response_

### Environment

```markdown
- Towhee version(e.g. v0.1.3 or 8b23a93):0.6.1
- OS(Ubuntu or CentOS):CentOS
- CPU/Memory:
- GPU: GeForce RTX 3090
- Others:
```

**Comment by binbinlv** [comment]
@reiase 
could you help to have a look here?

**Comment by reiase** [comment]
@junjiejiangjjj  It seems that the engine hangs when using GPU and multiprocessing. can you take a look at this issue?

**Comment by junjiejiangjjj** [comment]
It seems The CUDA runtime does not support the fork start method.   @angelapytao https://pytorch.org/docs/stable/notes/multiprocessing.html#cuda-in-multiprocessing

**Comment by junjiejiangjjj** [comment]
![image](https://user-images.githubusercontent.com/14136703/182060689-2e248878-4964-49e7-83c5-c4aa59e08d42.png)
One solution is to import torch in subprocess.

**Comment by angelapytao** [comment]
> 

I import torch in subproces. The issue still exist.

**Comment by junjiejiangjjj** [comment]
Could you share your code?

**Comment by angelapytao** [comment]
```
import os
import sys
from multiprocessing import Process, Queue, current_process, set_start_method
import numpy as np
from config import NPY_FILE
from logs import LOGGER
from urllib import request
# Create queues
from resnet50_encoder.encode import CustomOperator

task_queue = Queue(maxsize=10000)
done_queue = Queue()
NUMBER_OF_PROCESSES = 2
BATCH_SIZE = 2
GLOBAL_INIT = False
#
# Functions referenced by tasks
#

def image_embedding_resnet50(id, url):
    try:
        print("image_embedding_resnet50 come in")
        # image_encoder = CustomOperator()
        # print("init CustomOperator")
        # res = request.urlopen(url)
        # image_bytes = res.read()
        # embedding = image_encoder.execute(image_bytes)

        import torch
        from towhee import pipeline
        torch.device("cuda" if torch.cuda.is_available() else "cpu")
        pp = pipeline('towhee/image-embedding-resnet50')
        embedding = pp(url).tolist()
        print("after pipeline...")
        return id, embedding
    except Exception as e:
        LOGGER.error(f"Failed download img & extract embedding: {e}")
        print(f"Failed download img & extract embedding: {e}")
        return id, []

def write_npz_file(filename, ids, vecs):
    fname = os.path.join(NPY_FILE, filename)
    np.savez(fname, ids=ids, vectors=vecs)

# Submit tasks
def add_task(vid, img_path):
    global GLOBAL_INIT
    if not GLOBAL_INIT:
        do_task()
        GLOBAL_INIT = True
    try:
        tsk = (image_embedding_resnet50, (vid, img_path))
        task_queue.put(tsk)
        print(f"add_task id:{vid}")
    except Exception as e:
        print(f"InQueue error: {e}")
        return -1

def worker(input, output):
    print("worker start..")
    try:
        for func, args in iter(input.get, 'STOP'):
            print(f"iter(input.get) args:{args}")
            id, vec = calculate(func, args)
            print(f"calculate vec: {len(vec)}")
            if len(vec) == 2048:
                output.put([id, vec])
                print(f"worker put id{id}")
    except Exception as e:
        print(f"worker error: {e}")
        sys.exit(1)


#
# Function used to calculate result
#
def calculate(func, args):
    print("calculate come in, args1:", args[1])
    result = func(*args)
    print('%s says that %s%s = %s' % \
        (current_process().name, func.__name__, args, result))
    return result



def do_task():
    # Start worker processes
    try:
        for i in range(NUMBER_OF_PROCESSES):
            Process(target=worker, args=(task_queue, done_queue)).start()
        Process(target=pull_task, args=(done_queue,)).start()
    except Exception as e:
        LOGGER.error(f"Failed to embedding data: {e}")
    print("pull task start")

def do_write(start, idx, ids, embeddings):
    filename = "{}_{}.npz".format(start, idx)
    write_npz_file(filename, ids, embeddings)
    print("write {}, len: {} succ.".format(filename, len(ids)))

def pull_task(done_queue):
    # Get and print results
    try:
        ids = []
        embeddings = []
        start = 0
        idx = 0
        while True:
            try:
                # 队列为空，这里 get 会阻塞
                res = done_queue.get(block=True, timeout=10)
                if len(ids) == BATCH_SIZE:
                    do_write(start, idx, ids, embeddings)
                    start = idx + 1
                    ids = []
                    embeddings = []
                idx += 1
                ids.append(res[0])
                embeddings.append(res[1])
            except Exception as e:
                if len(ids) > 0:
                    do_write(start, idx, ids, embeddings)
                    start = idx + 1
                    ids = []
                    embeddings = []
    except Exception as e:
        LOGGER.error(f"Failed to exec task: {e}")
        # Tell child processes to stop
        for i in range(NUMBER_OF_PROCESSES):
            task_queue.put('STOP')
        sys.exit(1)

**Comment by junjiejiangjjj** [comment]
The process is blocked by queue, I change some code and the process can exit. It is just an easy fix to show you why the process is blocked.
![image](https://user-images.githubusercontent.com/14136703/182078779-5d1c9cc2-435d-41a3-a53b-9a43638448b6.png)
![image](https://user-images.githubusercontent.com/14136703/182078826-697db3a8-e673-49e1-b39c-39b0b17ab2d2.png)

I test it by the code:
![image](https://user-images.githubusercontent.com/14136703/182078922-e81071b8-6a50-4829-8bc2-9abded7ad4d8.png)

**Comment by angelapytao** [comment]
I changed ,and test 
<img width="1155" alt="image" src="https://user-images.githubusercontent.com/5369411/182080706-1ae1bca0-6016-4770-9e1e-df0ac615697a.png">

**Comment by angelapytao** [comment]
> ![image](https://user-images.githubusercontent.com/14136703/182078922-e81071b8-6a50-4829-8bc2-9abded7ad4d8.png)

The engine died and cant be restarted

**Comment by angelapytao** [comment]
And, I change tensorflow keras， it works OK in Centos

**Comment by angelapytao** [comment]
comment  the line  pp = pipeline('towhee/image-embedding-resnet50'), change another way to exatract embeddings, the whole code works fine. I think the reason caused by pipeline('towhee/image-embedding-resnet50')

**Comment by junjiejiangjjj** [comment]
Please use traceback to print the full stack 
![image](https://user-images.githubusercontent.com/14136703/182081780-c9d9665d-7759-4099-9be6-3bfdac0f8058.png)

**Comment by angelapytao** [comment]
2022-08-01 14:12:20,838 - 140562497996608 - save.py-save:32 - ERROR: Traceback (most recent call last):
  File "/angelapytao/milvus_web_server/milvus_api_server/operations/save.py", line 26, in image_embedding_resnet50
    pp = pipeline('towhee/image-embedding-resnet50')
  File "/angelapytao/milvus_web_server/milvus_api_server/env/lib/python3.9/site-packages/towhee/engine/factory.py", line 224, in pipeline
    start_engine()
  File "/angelapytao/milvus_web_server/milvus_api_server/env/lib/python3.9/site-packages/towhee/engine/engine.py", line 165, in start_engine
    raise Exception('The engine died and cant be restarted')
Exception: The engine died and cant be restarted

2022-08-01 14:12:20,838 ｜ ERROR ｜ save.py ｜ image_embedding_resnet50 ｜ 32 ｜ Traceback (most recent call last):
  File "/angelapytao/milvus_web_server/milvus_api_server/operations/save.py", line 26, in image_embedding_resnet50
    pp = pipeline('towhee/image-embedding-resnet50')
  File "/angelapytao/milvus_web_server/milvus_api_server/env/lib/python3.9/site-packages/towhee/engine/factory.py", line 224, in pipeline
    start_engine()
  File "/angelapytao/milvus_web_server/milvus_api_server/env/lib/python3.9/site-packages/towhee/engine/engine.py", line 165, in start_engine
    raise Exception('The engine died and cant be restarted')
Exception: The engine died and cant be restarted

**Comment by angelapytao** [comment]
And I use pipeline found another issue, when the python program works a while, the process hang on, the process exist, and has no exception. But client cannot access the interface anymore.
<img width="1204" alt="image" src="https://user-images.githubusercontent.com/5369411/182084631-599c8e94-f1af-4b67-a7ec-2f0b250a86c6.png">

**Comment by angelapytao** [comment]
There is no
 `_end_op ends with status: RunnerStatus.FINISHED, graph_context.py-graph_context:136 - INFO: Graph 25872 end.`, so I think why the program hangs on...

**Comment by junjiejiangjjj** [comment]
> Exception: The engine died and cant be restarted

Please find towhee/towhee/engine/task_scheduler.py, 
change the code of  `TaskScheduler.schedule_forever`

```python
    def schedule_forever(self, sleep_ms):
        """Runs the a single schedule step in a loop.

        sleep_ms: (`int`)
            Milliseconds to sleep after completing a single scheduling step.
        """
        while not self._need_stop:
            try:
                self.schedule_step()
                self._event.wait(sleep_ms / 1000)
            except Exception as e:
                import traceback
                print(e)
                print(traceback.print_exc())

```

**Comment by junjiejiangjjj** [comment]
> There is no `_end_op ends with status: RunnerStatus.FINISHED, graph_context.py-graph_context:136 - INFO: Graph 25872 end.`, so I think why the program hangs on...

Can you open another issue? And if the problem occurs again, please use `stack` or `pstrace` to find when the process hangs.

**Comment by angelapytao** [comment]
> > Exception: The engine died and cant be restarted
> 
> Please find towhee/towhee/engine/task_scheduler.py, change the code of `TaskScheduler.schedule_forever`
> 
> ```python
>     def schedule_forever(self, sleep_ms):
>         """Runs the a single schedule step in a loop.
> 
>         sleep_ms: (`int`)
>             Milliseconds to sleep after completing a single scheduling step.
>         """
>         while not self._need_stop:
>             try:
>                 self.schedule_step()
>                 self._event.wait(sleep_ms / 1000)
>             except Exception as e:
>                 import traceback
>                 print(e)
>                 print(traceback.print_exc())
> ```

<img width="1172" alt="image" src="https://user-images.githubusercontent.com/5369411/182094999-85802014-4ede-4030-a22e-f2f6570d6d23.png">

**Comment by junjiejiangjjj** [comment]
> > > Exception: The engine died and cant be restarted
> > 
> > 
> > Please find towhee/towhee/engine/task_scheduler.py, change the code of `TaskScheduler.schedule_forever`
> > ```python
> >     def schedule_forever(self, sleep_ms):
> >         """Runs the a single schedule step in a loop.
> > 
> >         sleep_ms: (`int`)
> >             Milliseconds to sleep after completing a single scheduling step.
> >         """
> >         while not self._need_stop:
> >             try:
> >                 self.schedule_step()
> >                 self._event.wait(sleep_ms / 1000)
> >             except Exception as e:
> >                 import traceback
> >                 print(e)
> >                 print(traceback.print_exc())
> > ```
> 
> <img alt="image" width="1172" src="https://user-images.githubusercontent.com/5369411/182094999-85802014-4ede-4030-a22e-f2f6570d6d23.png">

Are there any errors from `towhee/towhee/engine/task_scheduler.py::TaskScheduler.schedule_forever`
The error means the backend thread has exited, TaskScheduler.schedule_forever is the thread loop, we can catch the exception and find the errors.

**Comment by angelapytao** [comment]
No errors from towhee/towhee/engine/task_scheduler.py::TaskScheduler.schedule_forever

**Comment by angelapytao** [comment]
<img width="1234" alt="image" src="https://user-images.githubusercontent.com/5369411/182101332-8f616425-2fbb-4d69-8553-5312915357fb.png">
I change CustomOperator, It works fine.

**Comment by junjiejiangjjj** [comment]
In the parent process, does there create a pipeline?

**Comment by junjiejiangjjj** [comment]
In the parent process, does there create a pipeline?

**Comment by angelapytao** [comment]
> In the parent process, does there create a pipeline?

Yes, in main create a pipline for other interface to use.

**Comment by angelapytao** [comment]
<img width="1024" alt="image" src="https://user-images.githubusercontent.com/5369411/182116817-90500bfc-edae-4fc3-9dd1-97152838ac16.png">
<img width="565" alt="image" src="https://user-images.githubusercontent.com/5369411/182117207-ab106dad-6761-4bec-b3c3-a83d0fcaf82b.png">
I catch exception in sub process. Is the reason parent Empty exception influce the pipeline engine start?

**Comment by junjiejiangjjj** [comment]
> > In the parent process, does there create a pipeline?
> 
> Yes, in main create a pipline for other interface to use.

towhee pipeline would create a thread,  and in the subprocess, the thread died and could not start again.

**Comment by junjiejiangjjj** [comment]
Towhee has some backend threads, global variables, and locks,  if you want to start the subprocess to run pipeline,  use `Popen`, or just run the pipeline in other threads. @angelapytao

**Comment by angelapytao** [comment]
OK，got it. Thx!
